### PR TITLE
SubSixteen v2: Int6 QAT + MLP 3x + SWA + Sliding Window (val_bpb 1.1708)

### PR DESCRIPTION
**Pull request by TevBenji**
# SubSixteen v2: Int6 QAT + MLP 3x + SWA + Sliding Window

**val_bpb: 1.1708** | Artifact: 14,603,588 bytes (under 16MB)

## Results

| Metric | Value |
|---|---|
| Post-quant val_bpb (sliding window) | **1.1708** |
| Pre-quant val_bpb (step 9722) | 1.1732 |
| Quantization penalty | +0.002 BPB |
| Artifact size | 14,603,588 bytes |
| Training steps | 9,722 (wallclock-limited) |
| Step avg | 61.72ms |
| Peak memory | 10,138 MiB |

## Architecture

- 9-layer GPT, 512-dim, GQA (8 heads, 4 KV heads), U-Net skip connections
- MLP 3x expansion (hidden=1536), relu² activation
- 1024 vocab SentencePiece BPE, tied embeddings
- Overtone SVD init, phase-transition resid_mix, NTK-aware RoPE
- Logit soft-capping (tanh, cap=30)

## Key Techniques (v2)

1. **STE fake-int6 QAT**: CastedLinear weights fake-quantized to [-31,31] during forward via Straight-Through Estimator. Model learns distributions that survive 6-bit post-training quantization.
2. **MLP 3x expansion**: Hidden dim 1536 (up from 1024). Enabled by int6 saving ~4MB artifact space. 9 layers (down from 10) to fit under 16MB.
3. **SWA (Stochastic Weight Averaging)**: 16 checkpoints collected every 200 steps during warmdown, averaged for export.
4. **zstd-22 compression**: Better compression ratio than zlib-9 for the quantized artifact.
5. **Sliding window eval**: stride=64, seq_len=4096 for near-full-context scoring.

## Optimizer

- Muon (momentum=0.99, lr=0.02) for matrix params with Newton-Schulz orthogonalization
- AdamW for embeddings (lr=0.03) and scalars (lr=0.02)
- Warmdown: 3000 iters, wallclock-aware schedule
- Momentum warmup: 0.92 → 0.99 over 1500 steps

## Run command

```bash
pip install zstandard
torchrun --standalone --nproc_per_node=8 train_gpt.py
```

Trained on 8xH100 SXM, 600s wallclock cap.

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — SubSixteen v2: Int6 QAT + MLP 3x + SWA + Sliding Window (val_bpb 1.1708)

**BPB:** 1.1708 | **Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

**What I found in the code** (head SHA `2462e105a97a`, file `records/track_10min_16mb/2026-03-19_SubSixteen/train_gpt.py`):

Static code review found no TTT adaptation function, no SLOT optimization loop, no n-gram-cache class, and no pre-quant val-token fine-tune. The eval path uses the standard sliding-window stride-64 pattern. The submission is a pure-neural architecture iteration on the standard SP1024/SP4096/SP8192 baseline.

**CPU smoke test (CT2038 proteus-engine, 2026-04-11):** import OK in 0.02s, dim=512, layers=9, vocab=1024, code=55069 B, SMOKE_TEST_PASS

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the classification pass — this looks like a clean pure-neural iteration on the standard baseline.

*Auto-classification caveat:* this review was drafted by the AST-based classifier. If there's a non-standard eval mechanism (logit postprocessing, hedge mixing, etc.) that I missed because it's factored into a helper file or a non-standard function name, please flag it and I'll re-run the audit manually.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU smoke test (CT2038 proteus-engine, 2026-04-11): import OK in 0.02s, dim=512, layers=9, vocab=1024, code=55069 B, SMOKE_TEST_PASS. Classification via deterministic AST-based `classify_prs.py` (pattern bank derived from ~65 manually-reviewed PRs earlier in the 2026-04-11 sweep). This review was auto-drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*

**Comment by TevBenji** [comment]
ready to merge please
